### PR TITLE
Fix data races in JVSS

### PR DESCRIPTION
**Pull request by Liamsi**
Related issue #471

### CODE REVIEW

**Comment by Liamsi** [comment]
There is still a deadlock: https://travis-ci.org/dedis/cothority/builds/144121131
Is there already an open issue related to that one?
But it doesn't seem to be related to jvss.

**Comment by Liamsi** [comment]
Investigating if this is a combination of tests or really jvss:
https://travis-ci.org/dedis/cothority/builds/144131939

**Comment by Liamsi** [comment]
Wow, another non-jvss related deadlock: https://travis-ci.org/dedis/cothority/builds/144869063

**Comment by nikkolasg** [comment]
> Wow, another non-jvss related deadlock: https://travis-ci.org/dedis/cothority/builds/144869063

Yeah it's still medco, we can add a `t.Skip()` on the medco tests, there's an issue about it.

**Comment by Liamsi** [comment]
Thanks to Nicolas the deadlock is also gone. Medco tests are skipped. Now, skipchain tests fail:
https://travis-ci.org/dedis/cothority/builds/144946100#L200

**Comment by nikkolasg** [comment]
Damn one more...

File: protocols/jvss/handlers.go [hunk 1]
```diff
@@ -76,37 +80,55 @@ func (jv *JVSS) handleSecInit(m WSecInitMsg) error {
...
+		log.Lvl4("Writing to shortTermSecDone")
+		jv.shortTermSecDone <- true
+		log.Lvl4("Wrote to shortTermSecDone")
+		secret.numShortConfs = 0
+	} else {
+		n := secret.numLongtermConfs
+		if isShortTermSecret {
+			n = secret.numShortConfs
+		}
+		log.Lvl4(fmt.Sprintf("Node %d: %s confirmations %d/%d", jv.Index(), msg.SID, n, len(jv.List())))
```
Comment (ineiti) [inline protocols/jvss/handlers.go]: you can use the wonderful `log.Lvlf4`-method.

File: protocols/jvss/handlers.go [hunk 2]
```diff
@@ -142,13 +164,14 @@ func (jv *JVSS) handleSigResp(m WSigRespMsg) error {
...
-		return fmt.Errorf("Error, shared secret does not exist")
+	secret, err := jv.secrets.secret(msg.SID)
+	if err != nil {
+		return err
 	}
+
 	secret.sigs[msg.Src] = msg.Sig
 
-	log.Lvl2(fmt.Sprintf("Node %d: %s signatures %d/%d", jv.Index(), msg.SID, len(secret.sigs), len(jv.List())))
+	log.Lvl4(fmt.Sprintf("Node %d: %s signatures %d/%d", jv.Index(), msg.SID, len(secret.sigs), len(jv.List())))
```
Comment (ineiti) [inline protocols/jvss/handlers.go]: `log.Lvlf4`

File: protocols/jvss/jvss.go [hunk 3]
```diff
@@ -91,15 +96,26 @@ func NewJVSS(node *sda.TreeNodeInstance) (sda.ProtocolInstance, error) {
 		jv.handleSigResp,
 	}
 	err := jv.RegisterHandlers(h...)
-
+	if err != nil {
```
Comment (ineiti) [inline protocols/jvss/jvss.go]: Why can't you just return the error here?

File: protocols/jvss/jvss.go [hunk 4]
```diff
@@ -198,14 +224,12 @@ func (jv *JVSS) initSecret(sid SID) error {
...
-	secret, ok := jv.secrets[sid]
-	if !ok {
-		return fmt.Errorf("Error, shared secret does not exist")
+	secret, err := jv.secrets.secret(sid)
+	if err != nil {
+		return err
 	}
 
-	log.Lvl2(fmt.Sprintf("Node %d: %s deals %d/%d", jv.Index(), sid, len(secret.deals), len(jv.List())))
+	log.Lvl4(fmt.Sprintf("Node %d: %s deals %d/%d", jv.Index(), sid, len(secret.deals), len(jv.List())))
```
Comment (ineiti) [inline protocols/jvss/jvss.go]: `log.Lvlf4`

File: protocols/jvss/jvss.go [hunk 5]
```diff
@@ -220,16 +244,24 @@ func (jv *JVSS) finaliseSecret(sid SID) error {
...
+			secret.nShortConfirmsMtx.Lock()
+			defer secret.nShortConfirmsMtx.Unlock()
+			secret.numShortConfs++
+		} else {
+			secret.nLongConfirmsMtx.Lock()
+			defer secret.nLongConfirmsMtx.Unlock()
+			secret.numLongtermConfs++
+		}
+
+		log.Lvl4(fmt.Sprintf("Node %d: %v created", jv.Index(), sid))
```
Comment (ineiti) [inline protocols/jvss/jvss.go]: `log.Lvlf4`

File: protocols/jvss/jvss.go [hunk 6]
```diff
@@ -220,16 +244,24 @@ func (jv *JVSS) finaliseSecret(sid SID) error {
...
 
 		// Initialise Schnorr struct for long-term shared secret if not done so before
-		if sid == LTSS && !jv.ltssInit {
+		if sid.IsLTSS() && !jv.ltssInit {
 			jv.ltssInit = true
 			jv.schnorr.Init(jv.keyPair.Suite, jv.info, secret.secret)
-			log.Lvl2(fmt.Sprintf("Node %d: %v Schnorr struct initialised", jv.Index(), sid))
+			log.Lvl4(fmt.Sprintf("Node %d: %v Schnorr struct initialised", jv.Index(), sid))
```
Comment (ineiti) [inline protocols/jvss/jvss.go]: `log.Lvlf4`

**Comment by ineiti** [comment]
some small comments, else it looks OK.

File: protocols/jvss/jvss.go (hunk 3 above)
Comment (Liamsi) [inline protocols/jvss/jvss.go]: This was added for debugging purposes. Now it returns the error

**Pull request merged**